src/NUnitFramework/framework/Internal/Commands/FixturePerTestCaseCommand.cs

```diff
@@ -0,0 +1,71 @@
+// ***********************************************************************
+// Copyright (c) 2019 Charlie Poole, Rob Prouse
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System;
+using NUnit.Framework.Interfaces;
+
+namespace NUnit.Framework.Internal.Commands
+{
+    /// <summary>
+    /// ConstructFixtureCommand constructs the user test object if necessary.
+    /// </summary>
+    public class FixturePerTestCaseCommand : BeforeAndAfterTestCommand
+    {
+        /// <summary>
+        /// Handles the construction and disposement of a fixture per test case
+        /// </summary>
+        /// <param name="innerCommand">The inner command to which the command applies</param>
+        public FixturePerTestCaseCommand(TestCommand innerCommand)
+            : base(innerCommand)
+        {
+            TestSuite testSuite = null;
+
+            ITest currentTest = Test;
+            while (currentTest != null && testSuite == null)
+            {
+                testSuite = testSuite ?? currentTest as TestSuite;
+                currentTest = currentTest.Parent;
+            }
+
+            Guard.ArgumentValid(testSuite != null, "FixturePerTestCaseCommand must reference a TestSuite", nameof(innerCommand));
+
+            ITypeInfo typeInfo = testSuite.TypeInfo;
+
+            BeforeTest = (context) =>
+            {
+                if (typeInfo != null && !typeInfo.IsStaticClass && testSuite.Fixture == null)
```

Comment:
So, it turns out the reason I am having problems is that the behavior is inconsistent. I get different behavior for the tests on different runs and if I run the tests in the debugger or straight from the command line. In the debugger, the tests always pass, but on the command line they **usually** fail. It sounds like a classic threading problem, but I can't figure out what is going on because they always behave correctly in the debugger. 🤕 

From what I have managed to see in the debugger, I have a theory. In some of the test runs, the `testSuite.Fixture` is not null, so we don't actually construct the object below, instead we use the previously constructed one. I believe that this works fine single threaded, but when we are running in parallel, the first test sets the **one** `testSuite.Fixture` on line 60 below, then subsequent parallel tests don't enter this block because it is no longer `null`.

@avilv do you remember why you test if the Fixture is null and set it below? It also seems wrong to set it back to `null` in the `AfterTest` if it wasn't set here.